--Exercise 1--

Transforming Raw Noise into Strategic Foresight: The Impact of Data Analysis
At its core, data analysis is the systematic process of cleaning, transforming, modeling, and interpreting raw data to uncover meaningful patterns, draw informed conclusions, and support smarter decision-making. Rather than relying on intuition or legacy assumptions, data analysis provides an empirical framework for understanding what happened, why it happened, and what is likely to happen next.

Why Data Analysis Matters in Modern Contexts
We live in an era of unprecedented data velocity and volume. Businesses, governments, and research institutions generate terabytes of information daily from digital transactions, IoT sensors, social media, and customer interactions. However, raw data on its own is just noise. Data analysis provides the tools to unlock its actual value.

In the modern landscape, its importance boils down to three strategic advantages:

Evidence-Based Decision Making: It eliminates guesswork, allowing organizations to allocate resources and forecast outcomes based on concrete trends rather than intuition.

Operational Efficiency: By identifying bottlenecks, redundancies, and supply chain frictions in real time, organizations can cut costs and streamline workflows.

Agility & Competitive Advantage: In fast-moving markets, the ability to analyze changing consumer behaviors or economic shifts as they happen lets leaders pivot long before their competitors do.

3 Key Real-World Applications Today
1. Healthcare & Predictive Medicine
Modern healthcare relies heavily on data analysis to improve patient outcomes and operational logistics. By evaluating electronic health records (EHRs), genetic sequencing, and biometric feeds from wearable devices, medical teams can identify early warning signs of chronic diseases and design personalized treatment plans. Additionally, predictive models help hospital administrators forecast ER admission rates and optimize staff scheduling.

2. Financial Services & Real-Time Fraud Detection
In banking and fintech, machine learning algorithms analyze millions of global transactions per second to establish behavioral baselines for individual accounts. When an anomaly occurs—such as a sudden high-value purchase in a distant location—the system flags or blocks the transaction instantly to prevent fraud. Analysis also powers automated credit scoring, portfolio risk management, and algorithmic trading.

3. E-Commerce & Supply Chain Optimization
Retail giants use advanced recommendation engines to analyze user search history, past purchases, and peer demographics, serving personalized product suggestions that drive revenue. Behind the scenes, logistics teams analyze historical demand patterns, weather forecasts, and route conditions to optimize inventory levels across fulfillment centers and minimize delivery times.

Key Takeaway: Data analysis has evolved from a back-office reporting function into the primary engine of modern organizational strategy—turning passive information into proactive intelligence.

In [5]:
import pandas as pd

# Exercise 2
# Load the Sleep dataset using pandas
file_path = '/content/Time Americans Spend Sleeping.csv'
df = pd.read_csv(file_path)

# Show that it worked
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

print("=" * 60)
print("STEP 1: df.head() - First 5 rows")
print("=" * 60)

display(df.head())

print(f"Shape: {df.shape}")
print(f"  → {df.shape[0]} rows (samples/observations)")
print(f"  → {df.shape[1]} columns (features/variables)")

print("=" * 60)
print("STEP 3: df.columns - What variables do we have?")
print("=" * 60)

print(f"Columns ({len(df.columns)} total):")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

print("=" * 60)
print("STEP 4: df.info() - Data types & missing values")
print("=" * 60)

df.info()

print("=" * 60)
print("STEP 5: df.describe() - Statistical summary")
print("=" * 60)

display(df.describe())

Dataset loaded successfully!
Shape: 945 rows × 9 columns
STEP 1: df.head() - First 5 rows


,index,Year,Period,Avg hrs per day sleeping,Standard Error,Type of Days,Age Group,Activity,Sex
0,0,2003,Annual,8.57,0.018,All days,15 years and over,Sleeping,Both
1,1,2004,Annual,8.55,0.026,All days,15 years and over,Sleeping,Both
2,2,2005,Annual,8.62,0.023,All days,15 years and over,Sleeping,Both
3,3,2006,Annual,8.63,0.024,All days,15 years and over,Sleeping,Both
4,4,2007,Annual,8.57,0.024,All days,15 years and over,Sleeping,Both


Shape: (945, 9)
  → 945 rows (samples/observations)
  → 9 columns (features/variables)
STEP 3: df.columns - What variables do we have?
Columns (9 total):
  1. index
  2. Year
  3. Period
  4. Avg hrs per day sleeping
  5. Standard Error
  6. Type of Days
  7. Age Group
  8. Activity
  9. Sex
STEP 4: df.info() - Data types & missing values
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 945 entries, 0 to 944
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     945 non-null    int64  
 1   Year                      945 non-null    int64  
 2   Period                    945 non-null    object 
 3   Avg hrs per day sleeping  945 non-null    float64
 4   Standard Error            945 non-null    float64
 5   Type of Days              945 non-null    object 
 6   Age Group                 945 non-null    object 
 7   Activity                  945 non-null    object 
 8   

,index,Year,Avg hrs per day sleeping,Standard Error
count,945.000000,945.000000,945.000000,945.000000
mean,472.000000,2010.000000,8.806868,0.086958
std,272.942302,4.322782,0.545001,0.034631
min,0.000000,2003.000000,7.720000,0.018000
25%,236.000000,2006.000000,8.360000,0.063000
50%,472.000000,2010.000000,8.810000,0.084000
75%,708.000000,2014.000000,9.190000,0.106000
max,944.000000,2017.000000,10.650000,0.228000


### Dataset Description: Time Americans Spend Sleeping

This dataset contains survey data detailing the average number of hours Americans spend sleeping per day, categorized by various demographic and time-based factors.

**Key Features:**
*   **Timeframe:** Covers annual data from **2003 to 2017**.
*   **Demographics:** Includes data broken down by **Sex** (Men, Women, Both) and **Age Group** (e.g., 15 years and over, 65 years and over).
*   **Temporal Context:** Distinguishes between different **Types of Days**, such as weekends, holidays, and nonholiday weekdays.
*   **Metrics:** The primary value is **Avg hrs per day sleeping**, accompanied by the **Standard Error** to indicate measurement precision.
*   **Scope:** The data represents a total of **945 observations** across **9 variables**.

Exercise 3 - Data Types

index - Qualitative - This is an identifier code for each person surveyed.

Year - Quantiative - This represents the years over when the study was
conducted.

Period - Qualitative - This represents the period over which the subject was studied.

Avg hrs per day sleeping - Quantitative - This is the average number of days people slept during the survey.

Standard Error - Quantiative - This represents the statistical standard deviation error per subject.

Type of Days - Qualitative - This lists the type of days under which the survey was conducted.

Age Group - Qualitative - This is the categorical age groups of the survey.

Activity - Qualitative - This defines what the person was doing on that particular evening, whether they were sleeping or not.

Sex - Qualitative - This indicates whether men, women, or both sexes were being surveyed that day.

In [6]:
# Exercise 4
# Load the Sleep dataset using pandas
file_path = '/content/Iris.csv'
df = pd.read_csv(file_path)

# Show that it worked
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

print("=" * 60)
print("STEP 1: df.head() - First 5 rows")
print("=" * 60)

display(df.head())

print(f"Shape: {df.shape}")
print(f"  → {df.shape[0]} rows (samples/observations)")
print(f"  → {df.shape[1]} columns (features/variables)")

print("=" * 60)
print("STEP 3: df.columns - What variables do we have?")
print("=" * 60)

print(f"Columns ({len(df.columns)} total):")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

print("=" * 60)
print("STEP 4: df.info() - Data types & missing values")
print("=" * 60)

df.info()

print("=" * 60)
print("STEP 5: df.describe() - Statistical summary")
print("=" * 60)

display(df.describe())

Dataset loaded successfully!
Shape: 150 rows × 6 columns
STEP 1: df.head() - First 5 rows


,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


Shape: (150, 6)
  → 150 rows (samples/observations)
  → 6 columns (features/variables)
STEP 3: df.columns - What variables do we have?
Columns (6 total):
  1. Id
  2. SepalLengthCm
  3. SepalWidthCm
  4. PetalLengthCm
  5. PetalWidthCm
  6. Species
STEP 4: df.info() - Data types & missing values
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB
STEP 5: df.describe() - Statistical summary


,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,75.500000,5.843333,3.054000,3.758667,1.198667
std,43.445368,0.828066,0.433594,1.764420,0.763161
min,1.000000,4.300000,2.000000,1.000000,0.100000
25%,38.250000,5.100000,2.800000,1.600000,0.300000
50%,75.500000,5.800000,3.000000,4.350000,1.300000
75%,112.750000,6.400000,3.300000,5.100000,1.800000
max,150.000000,7.900000,4.400000,6.900000,2.500000


### Exercise 4 continued - Iris Dataset Data Types

Based on the inspection of `Iris.csv`, here is the classification of the columns:

**Quantitative (Numerical)**
*   **Id**: Quantitative (Discrete) - While it acts as a unique identifier, it is represented by numerical integers used to index the samples.
*   **SepalLengthCm**: Quantitative (Continuous) - Represents a physical measurement of the sepal length in centimeters.
*   **SepalWidthCm**: Quantitative (Continuous) - Represents a physical measurement of the sepal width in centimeters.
*   **PetalLengthCm**: Quantitative (Continuous) - Represents a physical measurement of the petal length in centimeters.
*   **PetalWidthCm**: Quantitative (Continuous) - Represents a physical measurement of the petal width in centimeters.

**Qualitative (Categorical)**
*   **Species**: Qualitative (Nominal) - This column categorizes the flowers into specific biological species (e.g., Iris-setosa, Iris-versicolor, Iris-virginica). It represents labels rather than numerical values.

In [7]:
#Exercise 5

import pandas as pd

# Exercise 2
# Load the Sleep dataset using pandas
file_path = '/content/Time Americans Spend Sleeping.csv'
df = pd.read_csv(file_path)

# Show that it worked
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

print("=" * 60)
print("STEP 1: df.head() - First 5 rows")
print("=" * 60)

display(df.head())

print(f"Shape: {df.shape}")
print(f"  → {df.shape[0]} rows (samples/observations)")
print(f"  → {df.shape[1]} columns (features/variables)")

print("=" * 60)
print("STEP 3: df.columns - What variables do we have?")
print("=" * 60)

print(f"Columns ({len(df.columns)} total):")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

print("=" * 60)
print("STEP 4: df.info() - Data types & missing values")
print("=" * 60)

df.info()

print("=" * 60)
print("STEP 5: df.describe() - Statistical summary")
print("=" * 60)

display(df.describe())

Dataset loaded successfully!
Shape: 945 rows × 9 columns
STEP 1: df.head() - First 5 rows


,index,Year,Period,Avg hrs per day sleeping,Standard Error,Type of Days,Age Group,Activity,Sex
0,0,2003,Annual,8.57,0.018,All days,15 years and over,Sleeping,Both
1,1,2004,Annual,8.55,0.026,All days,15 years and over,Sleeping,Both
2,2,2005,Annual,8.62,0.023,All days,15 years and over,Sleeping,Both
3,3,2006,Annual,8.63,0.024,All days,15 years and over,Sleeping,Both
4,4,2007,Annual,8.57,0.024,All days,15 years and over,Sleeping,Both


Shape: (945, 9)
  → 945 rows (samples/observations)
  → 9 columns (features/variables)
STEP 3: df.columns - What variables do we have?
Columns (9 total):
  1. index
  2. Year
  3. Period
  4. Avg hrs per day sleeping
  5. Standard Error
  6. Type of Days
  7. Age Group
  8. Activity
  9. Sex
STEP 4: df.info() - Data types & missing values
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 945 entries, 0 to 944
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     945 non-null    int64  
 1   Year                      945 non-null    int64  
 2   Period                    945 non-null    object 
 3   Avg hrs per day sleeping  945 non-null    float64
 4   Standard Error            945 non-null    float64
 5   Type of Days              945 non-null    object 
 6   Age Group                 945 non-null    object 
 7   Activity                  945 non-null    object 
 8   

,index,Year,Avg hrs per day sleeping,Standard Error
count,945.000000,945.000000,945.000000,945.000000
mean,472.000000,2010.000000,8.806868,0.086958
std,272.942302,4.322782,0.545001,0.034631
min,0.000000,2003.000000,7.720000,0.018000
25%,236.000000,2006.000000,8.360000,0.063000
50%,472.000000,2010.000000,8.810000,0.084000
75%,708.000000,2014.000000,9.190000,0.106000
max,944.000000,2017.000000,10.650000,0.228000


#Exercise 5 - Basic Observation Skills



#Analysis Strategy

Based on the variables in the **Time Americans Spend Sleeping** dataset, here are the columns most suitable for specific types of analysis:

#### 1. Trend Analysis (Over Time)
*   **Target Column:** `Year` and `Avg hrs per day sleeping`.
*   **Reasoning:** Since the data spans from 2003 to 2017, we can use the `Year` as a time-series axis to see if the average duration of sleep for Americans is increasing, decreasing, or staying stable over the decade.

#### 2. Group Comparison (Demographics)
*   **Target Columns:** `Sex` and `Age Group`.
*   **Reasoning:** These categorical columns allow us to compare sleep habits between different populations. For example, we could test if women sleep more than men on average, or if sleep duration significantly decreases as people move from the '15 years and over' category to older brackets.

#### 3. Behavioral Analysis (Contextual)
*   **Target Column:** `Type of Days`.
*   **Reasoning:** This is interesting for comparing sleep patterns on **Weekends vs. Weekdays**. It helps identify the "sleep debt" phenomenon where individuals might sleep significantly more on non-working days to compensate for the work week.

#Exercise 6 - Identifying Data Types

1- A company’s financial reports stored in an Excel file. -- Structured Data

2- Photographs uploaded to a social media platform - Unstructured Data

3- A collection of news articles on a website - Unstructured Data

4- Inventory data in a relational database - Structured Data

5- Recorded interviews from a market research study - Unstructured Data


#Exercise 7 - Transformation Exercise

For each of the following unstructured data sources, propose a method to convert it into structured data. Explain your reasoning.

1 - A series of blog posts about travel experiences.
2 - Audio recordings of customer service calls.
3 - Handwritten notes from a brainstorming session.
4 - A video tutorial on cooking.


### Exercise 7 - Proposed Data Transformations

To turn unstructured data into structured data, we typically extract key entities and metrics into a tabular format (rows and columns).

#### 1. Blog Posts about Travel
*   **Method:** Use **Natural Language Processing (NLP)** and **Named Entity Recognition (NER)** to extract specific fields.
*   **Structured Output:** A table with columns for *Destination City*, *Country*, *Travel Dates*, *Overall Rating (1-5)*, and *Keywords* (e.g., 'adventure', 'budget').
*   **Reasoning:** NLP can identify places and sentiments mentioned in text, allowing us to quantify the "experience" into searchable attributes.

#### 2. Audio Recordings of Customer Service Calls
*   **Method:** Use **Speech-to-Text (STT)** transcription followed by **Sentiment Analysis**.
*   **Structured Output:** A database containing *Call Duration*, *Customer ID*, *Agent Name*, *Reason for Call* (categorized), and *Sentiment Score* (Positive/Negative).
*   **Reasoning:** Converting audio to text allows software to count specific issues and track agent performance statistically.

#### 3. Handwritten Notes from a Brainstorming Session
*   **Method:** Use **Optical Character Recognition (OCR)** to digitize the text, then group by topic.
*   **Structured Output:** A spreadsheet with *Idea Category*, *Contributor Name*, *Priority Level*, and *Action Items*.
*   **Reasoning:** OCR extracts the raw text from images, which can then be organized into a project management tool for tracking tasks.

#### 4. Video Tutorial on Cooking
*   **Method:** **Video Analysis/Metadata Extraction** involving timestamps and computer vision.
*   **Structured Output:** A table listing *Ingredients*, *Quantities*, *Step Number*, *Cooking Time per Step*, and *Tools Required*.
*   **Reasoning:** Breaking a linear video into a step-by-step list makes it a recipe database that can be filtered by ingredients or total preparation time.

In [8]:
import pandas as pd

# Exercise 8
# Load the Sleep dataset using pandas
file_path = '/content/train.csv'
df = pd.read_csv(file_path)

# Show that it worked
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

print("=" * 60)
print("STEP 1: df.head() - First 5 rows")
print("=" * 60)

display(df.head())

print(f"Shape: {df.shape}")
print(f"  → {df.shape[0]} rows (samples/observations)")
print(f"  → {df.shape[1]} columns (features/variables)")

print("=" * 60)
print("STEP 3: df.columns - What variables do we have?")
print("=" * 60)

print(f"Columns ({len(df.columns)} total):")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

print("=" * 60)
print("STEP 4: df.info() - Data types & missing values")
print("=" * 60)

df.info()

print("=" * 60)
print("STEP 5: df.describe() - Statistical summary")
print("=" * 60)

display(df.describe())

Dataset loaded successfully!
Shape: 891 rows × 12 columns
STEP 1: df.head() - First 5 rows


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


Shape: (891, 12)
  → 891 rows (samples/observations)
  → 12 columns (features/variables)
STEP 3: df.columns - What variables do we have?
Columns (12 total):
  1. PassengerId
  2. Survived
  3. Pclass
  4. Name
  5. Sex
  6. Age
  7. SibSp
  8. Parch
  9. Ticket
  10. Fare
  11. Cabin
  12. Embarked
STEP 4: df.info() - Data types & missing values
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [10]:
# Exercise 9 - Export a dataframe to excel format and JSON format
import pandas as pd

# 1. Create a simple dataframe
data = {
    'Product': ['Laptop', 'Mouse', 'Monitor', 'Keyboard'],
    'Price': [1200, 25, 300, 75],
    'Quantity': [5, 50, 10, 20]
}
sample_df = pd.DataFrame(data)
print("Sample DataFrame created:")
display(sample_df)

# 2. Export the dataframe to an excel file
# Note: openpyxl is usually required for .xlsx files
excel_file = 'exported_data.xlsx'
sample_df.to_excel(excel_file, index=False, engine='openpyxl')
print(f"Successfully exported to {excel_file}")

# 3. Export the dataframe to a JSON file
json_file = 'exported_data.json'
sample_df.to_json(json_file, orient='records', indent=4)
print(f"Successfully exported to {json_file}")

Sample DataFrame created:


,Product,Price,Quantity
0,Laptop,1200,5
1,Mouse,25,50
2,Monitor,300,10
3,Keyboard,75,20


Successfully exported to exported_data.xlsx
Successfully exported to exported_data.json


In [13]:
# Exercise 10: Reading JSON Data

import pandas as pd

# Exercise 10
# Load the JSON dataset using pandas
file_path = '/content/posts.json'
# Corrected: Use read_json instead of read_csv
df = pd.read_json(file_path)

# Show that it worked
print(f"Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

print("=" * 60)
print("STEP 1: df.head() - First 5 rows")
print("=" * 60)

display(df.head())

print(f"Shape: {df.shape}")
print(f"  → {df.shape[0]} rows (samples/observations)")
print(f"  → {df.shape[1]} columns (features/variables)")

print("=" * 60)
print("STEP 3: df.columns - What variables do we have?")
print("=" * 60)

print(f"Columns ({len(df.columns)} total):")
for i, col in enumerate(df.columns, 1):
    print(f"  {i}. {col}")

print("=" * 60)
print("STEP 4: df.info() - Data types & missing values")
print("=" * 60)

df.info()

print("=" * 60)
print("STEP 5: df.describe() - Statistical summary")
print("=" * 60)

display(df.describe())

Dataset loaded successfully!
Shape: 100 rows × 4 columns
STEP 1: df.head() - First 5 rows


,userId,id,title,body
0,1,1,sunt aut facere repellat provident occaecati e...,quia et suscipit\nsuscipit recusandae consequu...
1,1,2,qui est esse,est rerum tempore vitae\nsequi sint nihil repr...
2,1,3,ea molestias quasi exercitationem repellat qui...,et iusto sed quo iure\nvoluptatem occaecati om...
3,1,4,eum et est occaecati,ullam et saepe reiciendis voluptatem adipisci\...
4,1,5,nesciunt quas odio,repudiandae veniam quaerat sunt sed\nalias aut...


Shape: (100, 4)
  → 100 rows (samples/observations)
  → 4 columns (features/variables)
STEP 3: df.columns - What variables do we have?
Columns (4 total):
  1. userId
  2. id
  3. title
  4. body
STEP 4: df.info() - Data types & missing values
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   userId  100 non-null    int64 
 1   id      100 non-null    int64 
 2   title   100 non-null    object
 3   body    100 non-null    object
dtypes: int64(2), object(2)
memory usage: 3.3+ KB
STEP 5: df.describe() - Statistical summary


,userId,id
count,100.000000,100.000000
mean,5.500000,50.500000
std,2.886751,29.011492
min,1.000000,1.000000
25%,3.000000,25.750000
50%,5.500000,50.500000
75%,8.000000,75.250000
max,10.000000,100.000000
